cns/restserver/ipam.go

```diff
@@ -122,132 +92,34 @@ func (service *HTTPRestService) releaseIPConfigHandler(w http.ResponseWriter, r
-			return true
-		}
-		return false
-	})
-	if len(existingIPConfigs) > 0 {
-		return fmt.Errorf("Failed to add IPConfigs to state, attempting to overwrite existing ipconfig states: %v", existingIPConfigs)
-	}
-
-	// add ipconfigs to state
-	for _, ipconfig = range ipconfigs {
-		if err = validateIPConfig(ipconfig); err != nil {
-			return err
-		}
-
-		service.PodIPConfigState[ipconfig.ID] = ipconfig
-		addedIPconfigs = append(addedIPconfigs, ipconfig)
-
-		if ipconfig.State == cns.Allocated {
-			var podInfo cns.KubernetesPodInfo
-
-			if err = json.Unmarshal(ipconfig.OrchestratorContext, &podInfo); err != nil {
-				return fmt.Errorf("Failed to add IPConfig to state: %+v with error: %v", ipconfig, err)
-			}
-
-			service.PodIPIDByOrchestratorContext[podInfo.GetOrchestratorContextKey()] = ipconfig.ID
-		}
-	}
-	return err
-}
-
-func filterIPConfigMap(toBeAdded map[string]cns.ContainerIPConfigState, f func(*cns.ContainerIPConfigState) bool) []*cns.ContainerIPConfigState {
-	vsf := make([]*cns.ContainerIPConfigState, 0)
-	for _, v := range toBeAdded {
-		if f(&v) {
-			vsf = append(vsf, &v)
-		}
-	}
-	return vsf
-}
-
-func (service *HTTPRestService) GetAllocatedIPConfigs() []*cns.ContainerIPConfigState {
+func (service *HTTPRestService) GetAllocatedIPConfigs() []*ipConfigurationStatus {
 	service.RLock()
 	defer service.RUnlock()
-	return filterIPConfigMap(service.PodIPConfigState, func(ipconfig *cns.ContainerIPConfigState) bool {
+	return filterIPConfigMap(service.PodIPConfigState, func(ipconfig *ipConfigurationStatus) bool {
 		return ipconfig.State == cns.Allocated
 	})
 }
 
-func (service *HTTPRestService) GetAvailableIPConfigs() []*cns.ContainerIPConfigState {
+func (service *HTTPRestService) GetAvailableIPConfigs() []*ipConfigurationStatus {
 	service.RLock()
 	defer service.RUnlock()
-	return filterIPConfigMap(service.PodIPConfigState, func(ipconfig *cns.ContainerIPConfigState) bool {
+	return filterIPConfigMap(service.PodIPConfigState, func(ipconfig *ipConfigurationStatus) bool {
 		return ipconfig.State == cns.Available
 	})
 }
 
-//RemoveIPConfigsFromState takes a lock on the service object, and will remove an array of ipconfigs to the CNS Service.
-//Used to add IPConfigs to the CNS pool, specifically in the scenario of rebatching.
-func (service *HTTPRestService) removeIPConfigsFromState(ipconfigs []cns.ContainerIPConfigState) error {
-	service.Lock()
-	defer service.Unlock()
-
-	for _, ipconfig := range ipconfigs {
-		delete(service.PodIPConfigState, ipconfig.ID)
-		var podInfo cns.KubernetesPodInfo
-		err := json.Unmarshal(ipconfig.OrchestratorContext, &podInfo)
-
-		// if batch delete failed return
-		if err != nil {
-			return err
+func filterIPConfigMap(toBeAdded map[string]ipConfigurationStatus, f func(*ipConfigurationStatus) bool) []*ipConfigurationStatus {
+	vsf := make([]*ipConfigurationStatus, 0)
+	for _, v := range toBeAdded {
+		if f(&v) {
+			vsf = append(vsf, &v)
```

Comment:
Does this need to return pointers? I believe there is a subtle bug here. The filtered list is being appended a pointer to the loop variable, which points to different values in each iteration. The result is that every item in the filtered list will actually be the same pointer, which will point to the last value added. An example here: https://play.golang.org/p/eE1HIR_KWmF

I would either change the function not to use pointers if possible. Otherwise, just take a copy of the variable and append a pointer to that copy.

Reply:
good catch 👍